### PR TITLE
Reusable widget for browsing QgsHistoryProvider logs

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR implements the next part of https://github.com/qgis/QGIS-Enhancement-Proposals/issues/130 , by implementing a reusable widget for browsing QgsHistoryEntries logged by history providers.

This hooks into the flexible QgsHistoryProviderRegistry mechanism, so that the same widget can be used in future for the Processing history and any plugin implemented history providers.

Here's the end result, shown browsing a proof-of-concept SQL query log:

https://user-images.githubusercontent.com/1829991/232976099-1199247a-29a0-46df-bb64-427340262292.mp4

### CODE REVIEW

File: src/gui/history/qgshistoryentrymodel.cpp [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+      // a bit of trickiness here, we need dates ordered descending
+      sortKey = QStringLiteral( "3: %1 %2" ).arg( QDate::currentDate().year() - timestamp.date().year(), 5, 10, QLatin1Char( '0' ) )
+                .arg( 12 - timestamp.date().month(), 2, 10, QLatin1Char( '0' ) );
+      groupString = timestamp.toString( QStringLiteral( "MMMM yyyy" ) );
+    }
+  }
+  return groupString;
+}
+
+QgsHistoryEntryDateGroupNode *QgsHistoryEntryRootNode::dateNode( const QDateTime &timestamp, QgsHistoryEntryModel *model )
```
Comment (nirvn) [inline src/gui/history/qgshistoryentrymodel.cpp:348]: Super nice to see this automatic handled by the history framework.

File: src/gui/history/qgshistoryentrynode.h [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+     * Returns the number of child nodes owned by this node.
+     */
+    virtual int childCount() const;
+
+    /**
+     * Returns a HTML formatted text string which should be shown to a user when
+     * selecting the node.
+     *
+     * Subclasses should implement this method or createWidget(), but not both.
+     */
```
Comment (nirvn) [inline src/gui/history/qgshistoryentrynode.h:75]: Maybe adding a \see createWidget() here (and a \see html() in the createWidget doc), it'll make it easier to jump back and forth in the documentation website.

File: src/gui/history/qgshistoryproviderregistry.h [hunk 3]
```diff
@@ -222,9 +182,34 @@ class GUI_EXPORT QgsHistoryProviderRegistry : public QObject
...
     bool clearHistory( Qgis::HistoryProviderBackend backend );
 
+  signals:
+
+    /**
+     * Emitted when an \a entry is added.
+     *
+     * \since QGIS 3.32
+     */
+    void entryAdded( long long id, const QgsHistoryEntry &entry, Qgis::HistoryProviderBackend backend );
```
Comment (nirvn) [inline src/gui/history/qgshistoryproviderregistry.h:197]: Seeking clarification on your implementation vision here: say I make a custom history provider in python for a plugin, what would the Qgis::HistoryProviderBackend value be here?  A plugin wouldn't be able to append a backend. Wouldn't it be better to tie this to an history provider ID instead?

File: src/gui/history/qgshistoryentry.cpp [hunk 4]
```diff
@@ -0,0 +1,37 @@
+/***************************************************************************
+                            qgshistoryentry.cpp
+                            -------------------------
+    begin                : April 2019
```
Comment (nirvn) [inline src/gui/history/qgshistoryentry.cpp:4]: Where you keeping this in store since 2019? :wink:

File: src/gui/history/qgshistoryentrynode.cpp [hunk 5]
```diff
@@ -0,0 +1,122 @@
+/***************************************************************************
+                            qgshistoryentrynode.cpp
+                            --------------------------
+    begin                : Aprial 2023
```
Comment (nirvn) [inline src/gui/history/qgshistoryentrynode.cpp:4]:
```suggestion
    begin                : April 2023
```

File: src/gui/history/qgshistoryentrynode.h [hunk 6]
```diff
@@ -0,0 +1,191 @@
+/***************************************************************************
+                            qgshistoryentrynode.h
+                            --------------------------
+    begin                : Aprial 2023
```
Comment (nirvn) [inline src/gui/history/qgshistoryentrynode.h:4]:
```suggestion
    begin                : April 2023
```

File: src/gui/history/qgshistoryproviderregistry.h (hunk 3 above)
Comment (nyalldawson) [reply src/gui/history/qgshistoryproviderregistry.h:197]: The backend is options like local, project, remote database etc. It defines where the history entry should be stored, whereas the provider is the "group" the action is associated with (eg processing algorithms, db queries, map exports, ....)

You can retrieve the provider id from the QgsHistoryEntry object.

Comment (nirvn) [reply src/gui/history/qgshistoryproviderregistry.h:197]: Should there be a plugins backend then?

Comment (nyalldawson) [reply src/gui/history/qgshistoryproviderregistry.h:197]: Plugins are free to use the existing backends. They just make up their own provider id and get all the history storage for free!

Comment (nirvn) [reply src/gui/history/qgshistoryproviderregistry.h:197]: Got it, I was misunderstanding backends as being categories themselves, not storage location / dataset. Thanks for the clarification.

File: src/gui/history/qgshistoryentry.cpp (hunk 4 above)
Comment (nyalldawson) [reply src/gui/history/qgshistoryentry.cpp:4]: I split it out from another file with that date... it seems right to keep the original :)

**Review by nirvn — APPROVED** [review]

**Pull request merged**